Adding an ellipsoid joint to the joint collection

Ipuch opened the pull request:
Thanks for taking the time to make and fill out this pull request! I'll be open to any enhancement. Thanks to @MegMll and @LucasJoseph for helping out on this one !

## Description
In this PR, we introduce the new ellipsoid joint to pinocchio. This implementation is based on the paper:

> Seth A, Sherman M, Eastman P, Delp S. Minimal formulation of joint motion for biomechanisms. Nonlinear Dyn. 2010 Oct 1;62(1):291-303. doi: 10.1007/s11071-010-9717-3. PMID: 21170173; PMCID: PMC3002261.

available [here](https://pmc.ncbi.nlm.nih.gov/articles/PMC3002261/pdf/nihms-199141.pdf). It has been adapted to the formalism of Pinocchio. It allows defining an ellipsoid on which a solid can glide and rotate with respect to the "normal" of the ellipsoid. The joint is driven as function of angles XYZ, meaningful for biomechanists and thoraco-scapular joint.

[Capture vidéo du 2025-11-06 23-35-48.webm](https://github.com/user-attachments/assets/e9ab1718-7ff5-4212-9a9e-8aad0e53a7cc)


<!--
Please include a clear and concise description of the changes included in this pull request.
Explain what are you changing and why.

If applicable, link the related issue using "Fixes #issue_number".
-->

## Checklist

- [x] I have run `pre-commit run --all-files` or `pixi run lint`
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary
- [x] I have made corresponding changes to the doxygen documentation
- [x] I have added tests that prove my fix or feature works
- [x] I have updated the [CHANGELOG](https://github.com/stack-of-tasks/pinocchio/blob/devel/CHANGELOG.md) or added the "no changelog" label if it's a CI-related issue
- [x] I have updated the [README credits section](https://github.com/stack-of-tasks/pinocchio?tab=readme-ov-file#credits)
- [x] We have run PINNOCHIO_EIGEN_NO_MALLOC

## Extra To-dos
- [x] **no reverse mode with ellipsoid joint for model-graph**
- [x] **Finite differences for motion subspace derivative Sdot**, as we have some non-zero terms in Sdot, we (with @MegMll) feel like it would be a good idea to have a test to check Sdot with finite differences as for S.

We have tested that the ellipsoid joint behave as a ZYX spherical joint when radii are zeros. We have to test the translation part as well with a composite joint.
- [x] **Rotation part** - The ellipsoid joint behaves as a ZYX spherical joint when radii are zeros. 
- [x] **Translation part** - Test the ellipsoid joint with a composite joint to check the translation part.
- [x] Add doc on **how to implement a joint** in `doc/e-development`

### Linting
- [x] pre-commit run --all-files

## Question
- [x] Should we add references **in the code** for what we have implemented ? Doc added in `joint-ellipsoid.hxx
- [x] To make it pass I had to modify second Dim and second MaxDim to 1 in this line : 
https://github.com/stack-of-tasks/pinocchio/blob/468acc3fbfbdf9052353fb703630ae70a4b2394e/include/pinocchio/multibody/joint-motion-subspace-generic.hpp#L52

- [x] **Model-graphs**, need to make sure the build of the graph happens the expected way for ellipsoid joint -> w**e can't reverse the joint ellipsoid** because the translations and the motion subspace are tight to the parent frame, and it would necessitate to have a new joint definition to get the transform in the child frame. To say it simply, we can't just modify `q` to express the spatial transform in the child frame.

Ipuch committed:
feat(joint-ellipsoid)

forcing subspace with one column

finite diff for S

Sdot

Ipuch committed:
refactor(joint-ellipsoid): extra methods for S and Sdot

Ipuch committed:
feat(joint-ellipsoid): bindings, model graphs

fix: ellipsoid radius variables, and linting

MegMll commented on bindings/python/parsers/graph/expose-edges.cpp:
```diff
@@ -111,6 +111,12 @@ namespace pinocchio
         .def_readonly("nv", &JointSphericalZYX::nv, "Number of tangent variables.");
 
+      bp::class_<JointEllipsoid>(
+        "JointEllipsoid", "Represents an ellipsoidal joint.",
```
Can you add the non default constructor, as well ?

Ipuch replied:
Done.

MegMll commented on bindings/python/parsers/graph/expose-edges.cpp:
Can you add nq and nv from the joint ellipsoid, to the collection at the top of the file ?

Ipuch replied:
Done.

MegMll commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,537 @@
+#include "pinocchio/math/matrix.hpp"
+#include "pinocchio/spatial/inertia.hpp"
+#include "pinocchio/spatial/skew.hpp"
+#include "pinocchio/multibody/joint-motion-subspace.hpp" // because we are dense.
```
Can you remove the comment, please ?

Ipuch replied:
Done.

MegMll commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,537 @@
+      const Scalar c1s2 = c1 * s2;
+
+      Scalar S_11, S_21, S_31, S_12, S_22, S_32;
+      // S_11 = c1 * (-radius_b * c0 * (c0 * s2 + c2 * s0 * s1) - radius_c * s0 * (- c0 * c2 * s1 +
```
Can you remove the commented lines, once you don't need them anymore, please ?

Ipuch replied:
Done.

MegMll commented on unittest/joint-ellipsoid.cpp:
```diff
@@ -0,0 +1,378 @@
+
+BOOST_AUTO_TEST_SUITE(JointEllipsoid)
+
+BOOST_AUTO_TEST_CASE(vsFreeFlyer)
```
There is no check here, so if you don't plan to use it, can you remove it, please .

Ipuch replied:
outdated

MegMll commented on unittest/joint-ellipsoid.cpp:
```diff
@@ -0,0 +1,378 @@
+  // Calculer jdata.S().transpose() * data.f[i]
+}
+
+// BOOST_AUTO_TEST_CASE(vsSphericalZYX)
```
Same, can you remove the commented test

Ipuch replied:
outdated.

MegMll commented on unittest/model-graph.cpp:
```diff
@@ -617,6 +617,66 @@ BOOST_AUTO_TEST_CASE(test_reverse_spherical_zyx)
     d_f.oMf[m_forward.getFrameId("body1", pinocchio::BODY)]));
 }
 
+/// @brief test if reversing of a composite joint is correct.
```
Can you change the doc over this test ?

Ipuch replied:
Done.

MegMll commented on unittest/joint-ellipsoid.cpp:
```diff
@@ -0,0 +1,378 @@
+    q_e(2) = std::atan2(-R(0, 1) / cy, R(0, 0) / cy); // z angle
+  }
+  else
+  {
+    // Gimbal lock - choose x = 0
+    q_e(0) = 0.0;
+    q_e(2) = std::atan2(R(1, 0), R(1, 1));
+  }
+
```
Could you use eigen function eulerAngles  ? Doc can be found here : https://libeigen.gitlab.io/eigen/docs-nightly/group__Geometry__Module.html#gaf67ac87150a889f5e320e39186b1c432

In eigen 5 it's now called canonicalEulerAngles, but in the previous version, which is mostly used with pinocchio, it was eulerAngles.

Ipuch replied:
Done.

MegMll commented on unittest/joint-ellipsoid.cpp:
```diff
@@ -0,0 +1,378 @@
+  std::cout << "✓ Spatial velocities match" << std::endl;
+
+  BOOST_CHECK(dataEllipsoid.oMi[1].isApprox(dataSphericalZYX.oMi[1]));
+  std::cout << "✓ Full oMi[1] matches" << std::endl;
```
Would it be possible to remove the std::cout ? because they appear even if the check does not work. If you want you can use, BOOST_CHECK_MESSAGE(), that will display a message if condition is wrong.

Ipuch replied:
Done. removed all messages "std::cout"

MegMll commented on unittest/model-graph.cpp:
```diff
@@ -617,6 +617,66 @@ BOOST_AUTO_TEST_CASE(test_reverse_spherical_zyx)
 }
 
+/// @brief test if reversing of a composite joint is correct.
+BOOST_AUTO_TEST_CASE(test_reverse_ellipsoid)
```
Nothing is checked yet in this test.

Ipuch replied:
Yes. I think I need help here to understand clearly what needs to be tested out @jorisv

Ipuch replied:
I need to test if I have the same pose by the modelgraph and by model.

Ipuch committed:
feat(joint-ellipsoid): extra utils for translations

position, velocities, accelerations

Ipuch committed:
tests(joint-ellipsoid): add composite test

delete extra commented test

linting(pre-comit)

tests(ellipsoid): enough core testing to me.

Ipuch committed:
tests(joint-ellipsoid): Sdot as finite difference

pre-commit

Ipuch committed:
test(joint-ellipsoid): model-graph

Ipuch committed:
docs(model-graphs)

Ipuch committed:
clean(joint-ellipsoid): unused variables and warnings

precomitted

Ipuch committed:
fix(joint-ellipsoid): binding python

not scalar but double instead

Ipuch committed:
feat(joint-ellipsoid): extra docs for binding + S available .

LucasJoseph committed:
Add ellipsoid joint kinematics example script

This script demonstrates the use of the ellipsoid joint in Pinocchio, showcasing kinematics, dynamics, and visualization of a robot with an ellipsoid joint. It includes functions for creating the robot model, computing kinematics and dynamics, and visualizing the motion.

Ipuch committed:
exemple(joint-ellipsoid)

Ipuch committed:
pre-commit

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+#include "pinocchio/math/matrix.hpp"
+#include "pinocchio/spatial/inertia.hpp"
+#include "pinocchio/spatial/skew.hpp"
+#include "pinocchio/multibody/joint-motion-subspace.hpp"
```
Check needed include (joint-motion-subspace is included two time)

Ipuch replied:
Done.

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+      data.S.matrix() << S_11, S_12, Scalar(0), S_21, S_22, Scalar(0), S_31, S_32, Scalar(0),
+        c1 * c2, s2, Scalar(0), -c1 * s2, c2, Scalar(0), s1, Scalar(0), Scalar(1);
+    }
+    void computeMotionSubspaceDerivative(
```
Why compute Sdot ?
It's maybe more effective to directly compute the bias

Ipuch replied:
moved to tests only

jorisv commented on include/pinocchio/multibody/joint-motion-subspace-generic.hpp:
```diff
@@ -49,7 +49,7 @@ namespace pinocchio
     typedef
       typename traits<JointMotionSubspaceTpl<Dim, Scalar, Options, MaxDim>>::DenseBase DenseBase;
-    typedef Eigen::Matrix<Scalar, Dim, Dim, Options, MaxDim, MaxDim> ReturnType;
+    typedef Eigen::Matrix<Scalar, Dim, 1, Options, MaxDim, 1> ReturnType;
```
I will verify it

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+    TangentVector_t joint_v;
+
+    Constraint_t S;
+    Constraint_t Sdot;
```
You can leave the Sdot computation in the unit test to validate the c computation

Ipuch replied:
done.

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+    {
+    }
+
+    explicit JointModelEllipsoidTpl(const Scalar & a, const Scalar & b, const Scalar & c)
```
```suggestion
    JointModelEllipsoidTpl(const Scalar & a, const Scalar & b, const Scalar & c)
```

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+  }; // struct JointDataEllipsoidTpl
+
+  PINOCCHIO_JOINT_CAST_TYPE_SPECIALIZATION(JointModelEllipsoidTpl);
+  template<typename _Scalar, int _Options>
```
Add doxygen comment to explain the joint behavior and the paper where it come from.
Also you can fill this page: doc/a-features/c-joints.md

Ipuch replied:
done.

jorisv replied:
Nice documentation, but it should be upside the model and not the data

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+      SINCOS(data.joint_q(2), &s2, &c2);
+
+      data.M.rotation() << c1 * c2, -c1 * s2, s1, c0 * s2 + c2 * s0 * s1, c0 * c2 - s0 * s1 * s2,
+        -c1 * s0, -c0 * c2 * s1 + s0 * s2, c0 * s1 * s2 + c2 * s0, c0 * c1;
```
You can deactivate clang-format to improve readability: see include/pinocchio/spatial/inertia.hpp:1008

```
      // clang-format off
      I_bar << dynamic_params[4], dynamic_params[5], dynamic_params[7],
               dynamic_params[5], dynamic_params[6], dynamic_params[8],
               dynamic_params[7], dynamic_params[8], dynamic_params[9];
      // clang-format on
```

Ipuch replied:
Done

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+      ny = -s0 * c1;
+      nz = c0 * c1;
+
+      data.M.translation() << radius_a * nx, radius_b * ny, radius_c * nz;
```
You can factorize the `data.M` computation

Ipuch replied:
Done

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+        s0, c0, s1, c1, s2, c2, dndotx_dqdot1, dndoty_dqdot0, dndoty_dqdot1, dndotz_dqdot0,
+        dndotz_dqdot1, data);
+
+      data.c.toVector().noalias() = data.Sdot.matrix() * data.joint_v;
```
You can optimize this computation by directly computing c.
Sdot computation is probably useless for end user, so you can put it in unit test to validate the c computation.

Ipuch replied:
Done

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+        radius_c
+        * (-c0 * c1 * q0dot * q0dot + s0 * s1 * q0dot * q1dot - s0 * c1 * q0ddot + s0 * s1 * q1dot * q0dot - c0 * c1 * q1dot * q1dot - c0 * s1 * q1ddot);
+      return a;
+    }
```
You can put this code in the unit test.

Ipuch replied:
Done

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+    template<typename ConfigVector, typename TangentVector>
+    Vector3 computeTranslationVelocities(
+      const typename Eigen::MatrixBase<ConfigVector> & qs,
+      const typename Eigen::MatrixBase<TangentVector> & vs) const
```
```suggestion
      const Eigen::MatrixBase<ConfigVector> & qs,
      const Eigen::MatrixBase<TangentVector> & vs) const
```

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+    }
+
+    template<typename ConfigVector>
+    Vector3 computeTranslations(const typename Eigen::MatrixBase<ConfigVector> & qs) const
```
```suggestion
    Vector3 computeTranslations(const Eigen::MatrixBase<ConfigVector> & qs) const
```

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+    Vector3 computeTranslationAccelerations(
+      const typename Eigen::MatrixBase<ConfigVector> & qs,
+      const typename Eigen::MatrixBase<TangentVector> & vs,
+      const typename Eigen::MatrixBase<TangentVector2> & as) const
```
```suggestion
      const Eigen::MatrixBase<ConfigVector> & qs,
      const Eigen::MatrixBase<TangentVector> & vs,
      const Eigen::MatrixBase<TangentVector2> & as) const
```

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+             + dndotz_dqdot1 * radius_c * c0 * c1;
+
+      data.S.matrix() << S_11, S_12, Scalar(0), S_21, S_22, Scalar(0), S_31, S_32, Scalar(0),
+        c1 * c2, s2, Scalar(0), -c1 * s2, c2, Scalar(0), s1, Scalar(0), Scalar(1);
```
You can deactivate clang-format to have a nice matrix

Ipuch replied:
Done

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,601 @@
+    };
+    typedef JointDataEllipsoidTpl<Scalar, Options> JointDataDerived;
+    typedef JointModelEllipsoidTpl<Scalar, Options> JointModelDerived;
+    typedef JointMotionSubspaceTpl<3, Scalar, Options, 3> Constraint_t;
```
You can specialize JointMotionSubspace to take advantage of S sparsity

Ipuch replied:
We will leave it for now.

jorisv commented on src/parsers/graph/model-graph-algo.cpp:
```diff
@@ -338,6 +345,28 @@ namespace pinocchio
 
+        void operator()(const JointEllipsoid & joint, const BodyFrame & b_f)
+        {
+          if (!edge.forward)
+            PINOCCHIO_THROW_PRETTY(
+              std::invalid_argument, "Graph - JointEllipsoid cannot be reversed yet.");
```
It could be nice to manage reverse joints

jorisv commented on unittest/joint-ellipsoid.cpp:
```diff
@@ -0,0 +1,356 @@
+    modelEllipsoid, dataEllipsoid, q_ellipsoid, qdot_ellipsoid, dataEllipsoid.tau,
+    Convention::WORLD);
+  BOOST_CHECK(aAbaEllipsoid.isApprox(qddot_ellipsoid));
+}
```
You can put the RNEA vs ABA test in a standalone test.

Ipuch replied:
Done.

Ipuch committed:
Update include/pinocchio/multibody/joint/joint-ellipsoid.hpp

Co-authored-by: Joris Vaillant <joris.vaillant@gmail.com>

Ipuch committed:
refactor(joint-ellipsoid): remove useless typename

Co-authored-by: Joris Vaillant <joris.vaillant@gmail.com>

Ipuch committed:
refactor(joint-ellipsoid): remove useless typenames

Co-authored-by: Joris Vaillant <joris.vaillant@gmail.com>

Ipuch committed:
refactor(joint-ellipsoid): remove useless typenames

Co-authored-by: Joris Vaillant <joris.vaillant@gmail.com>

Ipuch committed:
feat(joint-ellipsoid): computeBiais

LucasJoseph committed:
Factorized data.M computation
Preventing clang-format on large matrices

Ipuch committed:
fix(joint-ellispoid) : compute_biais

Ipuch committed:
refactor(joint-ellipsoid) : c instead of Sdot * qdot

LucasJoseph committed:
Move ComputeTranslations and ComputeTranslationVelocities to tests

LucasJoseph committed:
Moved computeTranslationAccelerations to tests

Ipuch committed:
refactor(joint-ellipsoid) : Sdot computeMotionSubspaceDerivative in tests only

y

Ipuch committed:
clean(joint-ellipsoid) no Sdot anymore.

Ipuch committed:
doc(joint-ellipsoid)

Ipuch committed:
tests(joint-ellipsoid): rnea vs aba isolated

Ipuch committed:
docs(joint-ellipsoid): changelog and readme

changelog again

Ipuch committed:
precommit

precommit again

Ipuch committed:
docs(joint-ellipsoid) - afeatures

scooting spherical ZYX

clean(joint-ellipsoid): extra include removed

hard precommit

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,555 @@
+    /// @param[in] c2, s2 Cosine and sine of q[2]
+    /// @param[out] data Joint data where M will be stored
+    void computeSpatialTransform(
+      Scalar c0, Scalar s0, Scalar c1, Scalar s1, Scalar c2, Scalar s2, JointDataDerived & data)
```
```suggestion
      const Scalar& c0, const Scalar& s0, const Scalar& c1, const Scalar& s1, const Scalar& c2, const Scalar& s2, JointDataDerived & data) const
```

This will avoid a copy when using scalar like casadi/cppad/cppadcg/boostmpfr

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,555 @@
+    JointModelEllipsoidTpl<NewScalar, Options> cast() const
+    {
+      typedef JointModelEllipsoidTpl<NewScalar, Options> ReturnType;
+      ReturnType res;
```
You have to forward all radius parameters to this copy

Ipuch replied:
done.

jorisv commented on README.md:
```diff
@@ -287,6 +287,7 @@ In addition to the core dev team, the following people have also been involved i
 -   [Sebastian Castro](https://roboticseabass.com) (RAI Institute): Viser visualizer and MeshCat visualizer feature extension
 -   [Lev Kozlov](https://github.com/lvjonok): Kinetic and potential energy regressors
 -   [Simeon Nedelchev](https://github.com/simeon-ned): Pseudo inertia and Log-Cholesky parametrization
+-   [Pierre Puchaud](https://github.com/ipuch) & [Lucas Joseph](https://github.com/LucasJoseph): Ellipsoid Joint
```
No need to share a line, there is room for everyone (You can add Lucas Joseph on a separate line)

Ipuch replied:
done.

jorisv commented on bindings/python/parsers/graph/expose-edges.cpp:
```diff
@@ -111,6 +114,25 @@ namespace pinocchio
+      bp::class_<JointEllipsoid>(
+        "JointEllipsoid", "Represents an ellipsoidal joint.",
+        bp::init<double, double, double>(
+          bp::args("self", "radius_a", "radius_b", "radius_c"), "Constructor with radii."))
```
```suggestion
      bp::class_<JointEllipsoid>(
        "JointEllipsoid", "Represents an ellipsoidal joint.",
        bp::init<>(bp::args("self"), "Default constructor."))
        .def_readonly("nq", &JointEllipsoid::nq, "Number of configuration variables.")
        .def_readonly("nv", &JointEllipsoid::nv, "Number of tangent variables.")
        .def(bp::init<double, double, double>(
          bp::args("self", "radius_a", "radius_b", "radius_c"), "Constructor with radii."))
```

jorisv commented on examples/ellipsoid-joint-kinematics.py:
You can add this file in the list `${PROJECT_NAME}_PYTHON_EXAMPLES` in `examples/CMakeLists.txt` to run it as an unit test.
This will ensure the script is run by the CI.

Ipuch replied:
done.

jorisv commented on src/parsers/graph/model-graph.cpp:
To support ModelGraph:
- Find a way to reverse the joint (we can modify input q and v to achieve that)
- Implement ReverseJointGraphVisitor
- Implement ReverseJointLimitsVisitor if needed
- Implement CreateJointModelVisitor
- Revert change on AddJointModelVisitor
- Implement UpdateJointGraphReversePoseVisitor if needed (if reversed joint parameters have changed)
- Implement ConfigurationConverterVisitor (check ZYX for inspiration)
- Implement TangentConverterVisitor
- Add the joint in `unittest/model-configuration-converter.cpp` unit test

Ipuch replied:
Aborted.

Ipuch committed:
Update include/pinocchio/multibody/joint/joint-ellipsoid.hpp

Co-authored-by: Joris Vaillant <joris.vaillant@gmail.com>

Ipuch committed:
Update bindings/python/parsers/graph/expose-edges.cpp

Co-authored-by: Joris Vaillant <joris.vaillant@gmail.com>

Ipuch committed:
review(@jorisv): docs and test

Ipuch committed:
scout(freeflyer docs)

Ipuch committed:
refactor(model-graph-algo): AddJointBetweenBodies

precommit

Ipuch committed:
doc(new joint): guide for developers.

jorisv committed:
doc: Fix doxygen doc parsing

jorisv committed:
core: Avoid cast build issue

jorisv committed:
python: Remove redundant member definition

jorisv committed:
lint: Run linter

jorisv committed:
examples: Try to overcome a Windows encoding issue

jorisv committed:
Revert "examples: Try to overcome a Windows encoding issue"

This reverts commit 36e708a9965fc76268456036c1a37e3d5c58302e.

jorisv committed:
ci: Another try to fix greek letter output issue

Ipuch committed:
feat/refactor(joint-ellipsoid): joint specialization

precommit

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,834 @@
+  template<typename Scalar, int Options, typename ForceSet>
+  struct ConstraintForceSetOp<JointMotionSubspaceEllipsoidTpl<Scalar, Options>, ForceSet>
+  {
+    typedef Eigen::Matrix<Scalar, 3, Eigen::Dynamic, Options> ReturnType;
```
This will do a dynamic allocation. You should use the standard definition:

```cpp
    typedef
      typename traits<JointMotionSubspaceEllipsoidTpl<Scalar, Options>>::DenseBase DenseBase;
    typedef
      typename MatrixMatrixProduct<Eigen::Transpose<const DenseBase>, ForceSet>::type ReturnType;
```

Ipuch replied:
Done.

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,834 @@
+    JointMotionSubspaceEllipsoidTpl()
+    {
+    }
+
```
```suggestion
```
No need of a default constructor if nothing happen inside.

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,834 @@
+      res.template block<5, 1>(0, 2).setZero();
+      res(5, 2) = Scalar(1);
+
+      return res;
```
```suggestion
      return S;
```

Ipuch replied:
Done.

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,834 @@
+      // w × e_z = [w[1], -w[0], 0]
+      res(ANGULAR + 0, 2) = w[1];
+      res(ANGULAR + 1, 2) = -w[0];
+      res(ANGULAR + 2, 2) = Scalar(0);
```
```suggestion
      res.col(2).noalias() = m.cross(SpatialAxis<5>());
```

This should do the same thing.
Like in revolute Joint you can create an Axis typedef: `typedef SpatialAxis<5> Axis`.

Ipuch replied:
Done. with an extra.

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,834 @@
+      // linear part: R^T * (t × e_z) = R^T * [-t_y, t_x, 0]^T
+      typedef Eigen::Matrix<Scalar, 3, 1, Options> Vector3;
+      const Vector3 t_cross_ez(-m.translation()[1], m.translation()[0], Scalar(0));
+      res.template block<3, 1>(LINEAR, 2).noalias() = m.rotation().transpose() * t_cross_ez;
```
```suggestion
      res.template block<3, 1>(LINEAR, 2).noalias() =
        m.rotation().transpose() * SpatialAxis<5>::CartesianAxis3::cross(m.translation());

```

Ipuch replied:
Done

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,834 @@
+
+        return res;
+      }
+    }; // struct TransposeConst
```
Don't do any optimization here. Do like JointMotionSubspaceTpl:

```cpp
      template<typename ForceSet>
      typename ConstraintForceSetOp<JointMotionSubspaceTpl, ForceSet>::ReturnType
      operator*(const Eigen::MatrixBase<ForceSet> & F)
      {
        return ref.S.transpose() * F.derived();
      }
```

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,834 @@
+    };
+    typedef JointDataEllipsoidTpl<Scalar, Options> JointDataDerived;
+    typedef JointModelEllipsoidTpl<Scalar, Options> JointModelDerived;
+    // typedef JointMotionSubspaceTpl<3, Scalar, Options, 3> Constraint_t;
```
```suggestion
```

Ipuch replied:
Done.

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,834 @@
+  {
+    // Y * S where Y is 6x6 inertia and S is 6x3 motion subspace
+    // Returns 6x3 force set
+    typedef InertiaTpl<S1, O1> Inertia;
```
```suggestion
```
Unused

Ipuch replied:
Finally, kept.

jorisv commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,834 @@
+    // Returns 6x3 force set
+    typedef InertiaTpl<S1, O1> Inertia;
+    Eigen::Matrix<S1, 6, 3, O1> M;
+    M.noalias() = Y.matrix() * S.matrix();
```
You can optimize this operation

Ipuch replied:
Done.

Ipuch committed:
refactor(joint-ellipsoid): review comments.

Co-authored-by: Joris Vaillant <joris.vaillant@gmail.com>

precommit

Ipuch commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,834 @@
+      res.setIndexes(id(), idx_q(), idx_v(), idx_vExtended());
+      return res;
+    }
+
```
Should it disappear ?

Ipuch commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,834 @@
+      const Scalar & c1,
+      const Scalar & s2,
+      const Scalar & c2,
+      const Scalar & dndotx_dqdot1,
+      const Scalar & dndoty_dqdot0,
+      const Scalar & dndoty_dqdot1,
+      const Scalar & dndotz_dqdot0,
+      const Scalar & dndotz_dqdot1,
+      JointDataDerived & data) const
+    {
+      Scalar S_11, S_21, S_31, S_12, S_22, S_32;
+      // clang-format off
+      S_11 = dndoty_dqdot0 * radius_b * (c0 * s2 + c2 * s0 * s1)
+             + dndotz_dqdot0 * radius_c * (-c0 * c2 * s1 + s0 * s2);
+
+      S_21 = -dndoty_dqdot0 * radius_b * (-c0 * c2 + s0 * s1 * s2)
+             + dndotz_dqdot0 * radius_c * (c0 * s1 * s2 + c2 * s0);
+
+      S_31 = c1 * (-dndoty_dqdot0 * radius_b * s0 + dndotz_dqdot0 * radius_c * c0);
+
+      S_12 = dndotx_dqdot1 * radius_a * c1 * c2
+             + dndoty_dqdot1 * radius_b * (c0 * s2 + c2 * s0 * s1)
+             + dndotz_dqdot1 * radius_c * (-c0 * c2 * s1 + s0 * s2);
+
+      S_22 = -dndotx_dqdot1 * radius_a * c1 * s2
+             - dndoty_dqdot1 * radius_b * (-c0 * c2 + s0 * s1 * s2)
+             + dndotz_dqdot1 * radius_c * (c0 * s1 * s2 + c2 * s0);
+
+      S_32 = dndotx_dqdot1 * radius_a * s1 - dndoty_dqdot1 * radius_b * c1 * s0
+             + dndotz_dqdot1 * radius_c * c0 * c1;
+
+      // Write directly to the internal matrix S, not to matrix() which returns a copy
+      data.S.S << S_11   , S_12  , Scalar(0),
+                  S_21   , S_22  , Scalar(0),
+                  S_31   , S_32  , Scalar(0),
+                  c1 * c2, s2    , Scalar(0),
+                 -c1 * s2, c2    , Scalar(0),
+                  s1     , Scalar(0), Scalar(1);
+      // clang-format on
+    }
```
Should it disappear then ? somehow ?

Ipuch commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,834 @@
+             + dndotz_dqdot1 * radius_c * c0 * c1;
+
+      // Write directly to the internal matrix S, not to matrix() which returns a copy
+      data.S.S << S_11   , S_12  , Scalar(0),
+                  S_21   , S_22  , Scalar(0),
+                  S_31   , S_32  , Scalar(0),
+                  c1 * c2, s2    , Scalar(0),
+                 -c1 * s2, c2    , Scalar(0),
+                  s1     , Scalar(0), Scalar(1);
```
I had to hack myself a bit.

Ipuch reviewed (commented):
One last problem with ConstraintForceSetOp which does not compile.

Ipuch committed:
fix(joint-ellipsoid): one more test fixed

Ipuch committed:
wontfix: test not passing

jorisv committed:
core: Fix build and tests

Ipuch committed:
clean(joint-ellipsoid): dead code and comments

precommit

jorisv committed:
core: Small improvments

jcarpent commented on bindings/python/parsers/graph/expose-edges.cpp:
```diff
@@ -111,6 +114,20 @@ namespace pinocchio
+        .def_readwrite(
+          "radius_b", &JointEllipsoid::radius_b, "Semi-axis length in the y direction.")
+        .def_readwrite(
+          "radius_c", &JointEllipsoid::radius_c, "Semi-axis length in the z direction.")
```
I would recommend changing the naming here. Something like `radius_x` or `semi_axis_x`. a, b, c is a bit meaningless.

Ipuch replied:
Set radius_xyz.

jcarpent commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
```diff
@@ -0,0 +1,817 @@
+    {
+      EIGEN_STATIC_ASSERT_VECTOR_SPECIFIC_SIZE(Vector3Like, 3);
+      // Compute first 6 rows from first 2 columns
+      Eigen::Matrix<Scalar, 6, 1> result = S.template block<6, 2>(0, 0) * v.template segment<2>(0);
```
```suggestion
      Eigen::Matrix<Scalar, 6, 1> result = S.template leftCols<2>() * v.template head<2>();
```

jcarpent commented on include/pinocchio/multibody/joint/joint-ellipsoid.hpp:
The default constructor should not have such values. It is better to set everything to 0 or 1.

Ipuch replied:
Set to zero

jcarpent commented on include/pinocchio/parsers/graph/joints.hpp:
```diff
@@ -161,6 +161,30 @@ namespace pinocchio
+    {
+      double radius_a = 0.01;
+      double radius_b = 0.01;
+      double radius_c = 0.01;
```
Default values should be changed?

Ipuch replied:
set to zero

jcarpent reviewed (changes requested):
Thanks @Ipuch @LucasJoseph and @MegMll for making this nice add-on to Pinocchio.
I made some minor remarks about member naming notably.

Ipuch committed:
Update include/pinocchio/multibody/joint/joint-ellipsoid.hpp

Co-authored-by: Justin Carpentier <justin.carpentier@inria.fr>

Ipuch committed:
review(joint-ellipsoid)

- radius_abc to radius_xyz
- default values to 0

precommit

jorisv committed:
core: Remove a codacy warning

jcarpent reviewed (approved):
All is green on my side.
Thank you all for this nice contribution.

Looking forward to more Biomechanics features @Ipuch @LucasJoseph ;)
